src/app/legacy/containers/CpsRecommendations/RecommendationsPromo/index.jsx

```diff
@@ -103,9 +104,29 @@ const StyledHeadline = styled.div`
 const RecommendationsPromo = ({ promo, eventTrackingData }) => {
   const { script, service } = useContext(ServiceContext);
   const handleClickTracking = useCombinedClickTrackerHandler(eventTrackingData);
-  const headline = pathOr(null, ['headlines', 'headline'], promo);
-  const url = pathOr(null, ['locators', 'assetUri'], promo);
-  const indexImage = pathOr(null, ['indexImage'], promo);
+  const headline =
+    pathOr(null, ['headlines', 'headline'], promo) ||
+    pathOr(
+      null,
+      [
+        'headlines',
+        'promoHeadline',
+        'blocks',
+        0,
+        'model',
+        'blocks',
+        0,
+        'model',
+        'text',
+      ],
+      promo,
+    );
+  const url =
+    pathOr(null, ['locators', 'assetUri'], promo) ||
+    pathOr(null, ['locators', 'canonicalUrl'], promo);
+  const indexImage =
+    pathOr(null, ['indexImage'], promo) ||
+    optimoToCPSImage({ cpsImage: pathOr(null, ['images'], promo) });
```

Comment:
This code here isn't explicit as to its intent, i.e. it processes the data differently depending on if it is a CPS recommendation or an Optimo recommendation. To be slightly more explicit and practice seperation of concerns you could create a new utility:
`const { headline, url, indexImage } = extractPromoData(promo);`

Inside that utility you could make it clear which logic is for CPS and which for Optimo e.g. 
```javascript
if(promo.locators.cpsUrn) {
  return {
     headline: pathOr(null, ['headlines', 'headline'], promo),
     ...
}

//optimo
return {
  headline: 
        pathOr(
      null,
      [
        'headlines',
        'promoHeadline',
        'blocks',
        0,
        'model',
        'blocks',
        0,
        'model',
        'text',
      ],
      promo,
    );
    ...
 }
 ```
 
 That way the `RecommendationsPromo` component itself doesn't need to know there is a difference between data formats for promos just that there is a seperate component that takes care of it. This is practicing the Open-Closed principle: https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle

Reply:
👍, When I got some time to spare I need to revise some of these fundamentals, and actually practice how to apply them and recognise any violations. 
This is now fixed